fair/forward.py

```diff
@@ -1,326 +1,309 @@
+             tcr_dbl=70.0,
+             F_volcanic=0.0,
+             F_solar=0.0,
+             aviNOx_frac=0.,
+             fossilCH4_frac=0.,
+             natural=np.array([202., 10.]),
+             useStevens=False,
+             efficacy=np.array([1.]*13),
+             scale=np.array([1.]*13),
+             oxCH4_frac=0.61,
+             lifetimes=False,
+             ghg_forcing="Etminan"):
+
+  # Conversion between ppm CO2 and GtC emissions
+  ppm_gtc   = M_ATMOS/1e18*molwt.C/molwt.AIR
+
+  # Conversion between ppb/ppt concentrations and Mt/kt emissions
+  # in the RCP databases ppb = Mt and ppt = kt so factor always 1e18
+  emis2conc = M_ATMOS/1e18*np.asarray(molwt.aslist)/molwt.AIR
+
+  # Funny units for nitrogen emissions - N2O is expressed in N2 equivalent
+  n2o_sf = molwt.N2O/molwt.N2
+  emis2conc[2] = emis2conc[2] / n2o_sf
+
+  # If TCR and ECS are supplied, calculate the q1 and q2 model coefficients 
+  # (overwriting any other q array that might have been supplied)
+  # ref eq. (4) and (5) of Millar et al ACP (2017)
+  k = 1.0 - (d/tcr_dbl)*(1.0 - np.exp(-tcr_dbl/d))  # Allow TCR to vary
   if type(tcrecs) in [np.ndarray,list]:
-    q =  (1.0 / F_2x) * (1.0/(k[0]-k[1])) \
-        * np.array([tcrecs[0]-k[1]*tcrecs[1],k[0]*tcrecs[1]-tcrecs[0]])
-
-  # # # ------------ SET UP OUTPUT TIMESERIES VARIABLES ------------ # # #
-  # the integ_len variable is used to store the length of our timeseries
-  # by default FAIR is not concentration driven
-  conc_driven=False
-  # here we check if FAIR is emissions driven
-  if type(emissions) in [np.ndarray,list]:
-    integ_len = len(emissions)
-    if (type(other_rf) in [np.ndarray,list]) and (len(other_rf)!=integ_len):
-        raise ValueError("The emissions and other_rf timeseries don't have the same length")
-    elif type(other_rf) in [int,float]:
-        other_rf = np.full(integ_len,other_rf)
-  
-  # here we check if FAIR is concentration driven
-  elif type(co2_concs) in [np.ndarray,list]:
-    integ_len = len(co2_concs)
-    conc_driven = True
-    if (type(other_rf) in [np.ndarray,list]) and (len(other_rf)!=integ_len):
-        raise ValueError("The concentrations and other_rf timeseries don't have the same length")
-    elif type(other_rf) in [int,float]:
-        other_rf = np.full(integ_len,other_rf)
-
-  # finally we check if only a non-CO2 radiative forcing timeseries has been supplied
-  elif type(other_rf) in [np.ndarray,list]:
-    integ_len = len(other_rf)
-    if type(emissions) in [int,float]:
-        emissions = np.full(integ_len,emissions)
+    q =  (1.0 / F2x) * (1.0/(k[0]-k[1])) * np.array([
+      tcrecs[0]-tcrecs[1]*k[1],tcrecs[1]*k[0]-tcrecs[0]])
+
+  # Convert any list to a numpy array for (a) speed and (b) consistency.
+  # Goes through all variables in scope and converts them.
+  frame = inspect.currentframe()
+  args, _, _, values = inspect.getargvalues(frame)
+  for i in args:
+    if type(values[i]) is list:
+      exec(i + '= np.array(' + i + ')')
+
```

Comment:
Clever but dangerous if we ever want to use lists of strings for something. Could we not just be more careful with making sure things are initialised as arrays?

Reply:
Could either 1. add exceptions for lists that we know will be strings, or 2. iterate through the numeric arrays and perform the check. If 1 then no action required at present as there should be no non-numeric arrays, but this is likely to change if we use pandas/xarray